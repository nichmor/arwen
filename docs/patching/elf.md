## ELF (Executable and Linkable Format)
`ELF` is the standard binary format used on Linux, BSD variants, Solaris, and many other Unix-like operating systems for `executables`, `shared libraries (.so files)`, `object code`, and `core dumps`.

- **Executables**: These are files containing machine code instructions that the operating system can load directly into memory and run as a program. They are the result of compiling and linking source code. On Linux, common executables like `/bin/bash` or `/usr/bin/grep` are typically in ELF format.

- **Shared Libraries (.so files)**: Standing for `"Shared Object"`, these files contain compiled code and data (functions, variables) designed to be used by multiple executable programs at the same time while they are running `(this is dynamic linking)`. Instead of each program having its own copy of the library code, they share a single copy loaded in memory. Examples include `libc.so.6` (the standard C library) or `libssl.so` (for SSL/TLS functions).

- **Object Code (.o files)**: This is the intermediate output produced by a compiler when it translates source code (like `.c` or `.rs` files) into machine code for a specific architecture. Object files aren't runnable on their own. They contain the compiled code and data but also information needed by the linker to combine them with other object files and libraries to create a final executable or shared library.

- **Core Dumps**: This is a file created by the operating system when a program crashes or terminates unexpectedly. It contains a snapshot of the program's state at the time of the crash, including its memory contents (the "core") and often CPU register values. Developers use core dump files with debuggers (like `gdb`) to perform post-mortem analysis and figure out why the program failed.

## ELF File Structure Overview


- **ELF Header**: Always located at the very beginning of the file (offset 0).

- **Program Header Table (PHT)**: Optional, but present in executables and shared libraries. Describes segments used for loading the file into memory. Its location and size are specified in the ELF Header.

- **Sections**: Contain the actual code, data, symbol tables, string tables, and linking information. Their locations and sizes are described by the `Section Header Table`.

- **Section Header Table (SHT)**: Present except sometimes in stripped executables. Describes the file's sections. Its location and size are specified in the ELF Header, often placed near the end of the file.

A simple ELF file structure looks like this:

```
+---------------------+
| ELF Header          |  (File Offset 0)
+---------------------+
| Program Header Table|  (Offset specified in ELF Header)
| (Optional)          |
+---------------------+
|                     |
| Segment 1 / Section |  (e.g., Code .text)
|                     |
+---------------------+
|                     |
| Segment 2 / Section |  (e.g., Data .data, .bss)
|                     |
+---------------------+
| ...                 |  (Other segments/sections like .dynamic)
+---------------------+
|                     |
| Section Header Table|  (Offset specified in ELF Header, often near end)
| (Optional)          |
+---------------------+
```

!!! note
    Sections are often contained within Segments.

    The PHT describes the file from a loading/memory perspective (Segments), while the SHT describes it from a linking/content perspective (Sections).


### More detailed view of the components

- **`1. ELF Header (Elf64_Ehdr / Elf32_Ehdr)`**
  - **Location:** Start of the file (offset 0).
  - **Purpose:** Identifies the file as ELF and provides essential metadata and pointers.
  - **Key Fields:**
    - `e_ident`: Magic number (`\x7fELF`) and other info (class 32/64-bit, data encoding, ABI version).
    - `e_type`: File type (e.g., `ET_EXEC` for executable, `ET_DYN` for shared library/position-independent executable, `ET_REL` for relocatable object file).
    - `e_machine`: Target architecture (e.g., `EM_X86_64`, `EM_AARCH64`).
    - `e_version`: ELF version (usually 1).
    - `e_entry`: Virtual memory address of the program's entry point (where execution begins).
    - `e_phoff`: File offset to the start of the Program Header Table.
    - `e_shoff`: File offset to the start of the Section Header Table.
    - `e_flags`: Processor-specific flags.
    - `e_ehsize`: Size of this ELF header.
    - `e_phentsize`: Size of a single entry in the Program Header Table.
    - `e_phnum`: Number of entries in the Program Header Table.
    - `e_shentsize`: Size of a single entry in the Section Header Table.
    - `e_shnum`: Number of entries in the Section Header Table.
    - `e_shstrndx`: Section header table index of the section containing section names.


- **2. Program Header Table (PHT)**
  -  **Location:** At the file offset specified by `e_phoff` in the ELF Header. It's an array of `e_phnum` entries, each `e_phentsize` bytes long.
  - **Purpose:** Describes *segments* â€“ contiguous chunks of the file that need to be mapped into memory by the system loader when creating a process image. This is the "execution view" of the file.
  -  **`Key Entry Types (p_type):`**
     - `PT_LOAD`: Describes a loadable segment (e.g., code, data). Specifies file offset (`p_offset`), virtual address (`p_vaddr`), physical address (`p_paddr` - often ignored), file size (`p_filesz`), memory size (`p_memsz` - can be larger for `.bss`), and permissions (`p_flags` - Read/Write/Execute).
     - `PT_DYNAMIC`: Points to the segment containing dynamic linking information (the `.dynamic` section). Specifies offset (`p_offset`) and size (`p_filesz`). Essential for executables/libraries using dynamic linking.
     - `PT_INTERP`: Points to a null-terminated string specifying the path of the program interpreter (dynamic linker, e.g., `/lib64/ld-linux-x86-64.so.2`).
     - Other types exist for notes (`PT_NOTE`), TLS (`PT_TLS`), etc.

- **3. Sections**
  - **Location:** Scattered throughout the file, as defined by the Section Header Table.
  - **Purpose:** Hold the actual content: compiled code, data, symbol tables, string tables, relocation information, dynamic linking structures, debugging info, etc. Sections represent the "linking view" of the file.
  - **Common Sections:**
  - `.text`: Executable code.
    - `.data`: Initialized data (global/static variables with initial values).
    - `.bss`: Uninitialized data (global/static variables without explicit initial values; occupies no file space but reserves memory space).
    - `.rodata`: Read-only data (constants, string literals).
    - `.symtab`: Symbol table (for linking/debugging).
    - `.strtab`: String table for `.symtab`.
    - `.shstrtab`: String table for section names themselves.
    - `.dynamic`: Holds the array of dynamic linking tags (see below).
    - `.dynsym`: Minimal symbol table needed for dynamic linking.
    - `.dynstr`: String table for `.dynsym` and `.dynamic` entries requiring strings (like library names in `DT_NEEDED`).
    - `.interp`: Contains the path string for the program interpreter (pointed to by `PT_INTERP`).


- **4. Section Header Table (SHT)**
  - **Location:** At the file offset specified by `e_shoff` in the ELF Header. An array of `e_shnum` entries, each `e_shentsize` bytes long.
  - **Purpose:** Describes each section in the file. Essential for linkers and debuggers, less so for the runtime loader (which uses the PHT).
  - **`Key Fields in an Entry (Elf64_Shdr / Elf32_Shdr):`**
    - `sh_name`: Offset into the `.shstrtab` section giving the section's name.
    - `sh_type`: Section type (e.g., `SHT_PROGBITS` for code/data, `SHT_SYMTAB` for symbols, `SHT_STRTAB` for strings, `SHT_NOBITS` for `.bss`, `SHT_DYNAMIC` for dynamic tags).
    - `sh_flags`: Attributes like `SHF_WRITE`, `SHF_ALLOC` (occupies memory), `SHF_EXECINSTR` (executable code).
    - `sh_addr`: Virtual memory address if the section is loaded.
    - `sh_offset`: File offset of the section's start.
    - `sh_size`: Size of the section in the file.
    - `sh_link`, `sh_info`: Interpretation depends on section type (e.g., for `.dynamic`, `sh_link` points to the string table `.dynstr`).
    - `sh_addralign`: Required alignment.
    - `sh_entsize`: Size of entries if the section holds a table (like symbol table).


## Dynamic Linking
For dynamic linking, the crucial parts are:

1.  The `PT_INTERP` program header entry (and the `.interp` section it points to) tells the kernel which dynamic linker to execute.
2.  The `PT_DYNAMIC` program header entry points to the segment containing the `.dynamic` section.
3.  The **`.dynamic`** section contains the array of tags and values that drive the dynamic linker. `arwen` modifies values associated with specific tags within this section (or the string table `.dynstr` they point to):
    * **`DT_INTERP`**: (Tag only, value points into `.dynstr`) Path to the dynamic linker. Modifying the string in `.dynstr` changes the interpreter.
    * **`DT_NEEDED`**: (Value points into `.dynstr`) Name of a required library. Modifying the string in `.dynstr` changes the dependency. `arwen` can also add/remove entries in the `.dynamic` array itself.
    * **`DT_RPATH`** / **`DT_RUNPATH`**: (Value points into `.dynstr`) Library search paths. Modifying the string in `.dynstr` changes these paths. Remember `$ORIGIN` is expanded by the linker to the directory of the object being processed.

**How the Dynamic Linker Uses This Information (Simplified version)**

The dynamic linker (`ld-linux.so.2` or similar), specified by `PT_INTERP` / `DT_INTERP`, reads the `.dynamic` section (found via `PT_DYNAMIC`). It processes `DT_NEEDED` entries to find required libraries, searching in paths derived from `DT_RUNPATH`, `LD_LIBRARY_PATH`, `DT_RPATH`, and system defaults, then loads them into memory (using their own ELF structures) and resolves symbols.

Understanding this workflow, helps understanding how `arwen` modifies an RPATH or a needed library. It's mostly targeting specific entries within the `.dynamic` section or the associated `.dynstr` string table within the ELF file layout.
