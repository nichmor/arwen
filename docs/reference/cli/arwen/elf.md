<!--- This file is autogenerated. Do not edit manually! -->
# <code>[arwen](../arwen.md) elf</code>

## About
ELF commands

--8<-- "docs/reference/cli/arwen/elf_extender.md:description"

## Usage
```
arwen elf <COMMAND>
```

## Subcommands
| Command | Description |
|---------|-------------|
| [`add-rpath`](elf/add-rpath.md) | Add a run path to the elf file |
| [`remove-rpath`](elf/remove-rpath.md) | Remove runpath from the elf file |
| [`set-rpath`](elf/set-rpath.md) | Set the rpath of the elf file |
| [`print-rpath`](elf/print-rpath.md) | Print the rpath of the ELF file |
| [`force-rpath`](elf/force-rpath.md) | Force usage of `DT_RPATH` instead of `DT_RUNPATH` |
| [`set-interpreter`](elf/set-interpreter.md) | Set the interpreter of the elf file |
| [`print-interpreter`](elf/print-interpreter.md) | Print the interpreter of the ELF file |
| [`set-os-abi`](elf/set-os-abi.md) | Change the OS ABI of the elf file |
| [`print-os-abi`](elf/print-os-abi.md) | Print the OS ABI of the elf file |
| [`set-soname`](elf/set-soname.md) | Sets the DT_SONAME field in the ELF file |
| [`print-soname`](elf/print-soname.md) | Print the `DT_SONAME` |
| [`shrink-rpath`](elf/shrink-rpath.md) | Remove from the DT_RUNPATH or DT_RPATH all directories that do not contain a library referenced by DT_NEEDED fields of the executable or library |
| [`add-needed`](elf/add-needed.md) | Add dependencies on a dynamic library  DT_NEEDED |
| [`remove-needed`](elf/remove-needed.md) | Remove dynamic library dependencies from DT_NEEDED |
| [`replace-needed`](elf/replace-needed.md) | Replace dependencies from DT_NEEDED |
| [`print-needed`](elf/print-needed.md) | Print the DT_NEEDED |
| [`no-default-lib`](elf/no-default-lib.md) | Disable the default library search paths in the elf file |
| [`clear-symbol-version`](elf/clear-symbol-version.md) | Clear a version for a specific symbol |
| [`rename-dynamic-symbols`](elf/rename-dynamic-symbols.md) | Renames dynamic symbols |
| [`add-debug-tag`](elf/add-debug-tag.md) | Adds `DT_DEBUG` tag to the elf file to enable debugging |
| [`clear-exec-stack`](elf/clear-exec-stack.md) | Clears the executable flag of the GNU_STACK program header |
| [`set-exec-stack`](elf/set-exec-stack.md) | Set the executable flag of the GNU_STACK program header |
| [`print-exec-stack`](elf/print-exec-stack.md) | Print the exec stack of the ELF file |


--8<-- "docs/reference/cli/arwen/elf_extender.md:example"
