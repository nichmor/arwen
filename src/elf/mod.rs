pub mod container;
pub mod rewriter;

pub use container::*;
pub use rewriter::*;
