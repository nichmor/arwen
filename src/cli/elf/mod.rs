use crate::elf::ElfError;

use super::ElfCommand;

pub mod add_debug_tag;
pub mod add_needed;
pub mod add_rpath;
pub mod clear_execstack;
pub mod clear_version_symbol;
pub mod force_rpath;
pub mod no_default_lib;
pub mod print_execstack;
pub mod print_interpreter;
pub mod print_needed;
pub mod print_os_abi;
pub mod print_rpath;
pub mod print_soname;
pub mod remove_needed;
pub mod remove_rpath;
pub mod rename_dynamic_symbols;
pub mod replace_needed;
pub mod set_execstack;
pub mod set_interpreter;
pub mod set_os_abi;
pub mod set_rpath;
pub mod set_soname;
pub mod shrink_rpath;

pub fn execute(elf: ElfCommand) -> Result<(), ElfError> {
    match elf {
        ElfCommand::AddRpath(args) => add_rpath::execute(args),
        ElfCommand::RemoveRpath(args) => remove_rpath::execute(args),
        ElfCommand::SetRpath(args) => set_rpath::execute(args),
        ElfCommand::ForceRpath(args) => force_rpath::execute(args),
        ElfCommand::PrintRpath(args) => print_rpath::execute(args),
        ElfCommand::SetInterpreter(args) => set_interpreter::execute(args),
        ElfCommand::PrintInterpreter(args) => print_interpreter::execute(args),
        ElfCommand::SetOsAbi(args) => set_os_abi::execute(args),
        ElfCommand::PrintOsAbi(args) => print_os_abi::execute(args),
        ElfCommand::SetSoname(args) => set_soname::execute(args),
        ElfCommand::PrintSoname(args) => print_soname::execute(args),
        ElfCommand::ShrinkRpath(args) => shrink_rpath::execute(args),
        ElfCommand::AddNeeded(args) => add_needed::execute(args),
        ElfCommand::RemoveNeeded(args) => remove_needed::execute(args),
        ElfCommand::ReplaceNeeded(args) => replace_needed::execute(args),
        ElfCommand::PrintNeeded(args) => print_needed::execute(args),
        ElfCommand::NoDefaultLib(args) => no_default_lib::execute(args),
        ElfCommand::ClearSymbolVersion(args) => clear_version_symbol::execute(args),
        ElfCommand::RenameDynamicSymbols(args) => rename_dynamic_symbols::execute(args),
        ElfCommand::AddDebugTag(args) => add_debug_tag::execute(args),
        ElfCommand::ClearExecStack(args) => clear_execstack::execute(args),
        ElfCommand::SetExecStack(args) => set_execstack::execute(args),
        ElfCommand::PrintExecStack(args) => print_execstack::execute(args),
    }
}
