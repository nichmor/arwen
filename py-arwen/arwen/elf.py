"""
ELF binary manipulation module for Linux/Unix systems.

This module provides functionality for working with ELF binary files,
allowing operations like modifying runpaths, changing interpreters, and more.
"""

from typing import Optional, Dict, List


from arwen.arwen import PyElfContainer


class ElfContainer:
    """
    Python wrapper for manipulating ELF binary files.

    This class provides methods to modify various aspects of ELF binaries,
    such as runpaths, needed libraries, interpreter paths, and more.
    """

    _inner: PyElfContainer

    @staticmethod
    def from_path(path: str) -> "ElfContainer":
        """
        Create a new ElfContainer by parsing an ELF binary file.

        Args:
            path: Path to the ELF binary file

        Returns:
            A new ElfContainer instance

        Raises:
            ElfError: If the file cannot be parsed as an ELF binary
            Exception: If the file cannot be read
        """
        return ElfContainer._from_py(PyElfContainer.from_path(path))

    def add_runpath(self, new_runpath: str) -> None:
        """
        Add a new runpath to the ELF file.

        Args:
            new_runpath: The runpath to add

        Raises:
            ElfError: If the runpath cannot be added
        """
        self._inner.add_runpath(new_runpath)

    def set_runpath(self, set_runpath: str) -> None:
        """
        Set runpath in the ELF file.

        Args:
            set_runpath: The runpath to set

        Raises:
            ElfError: If the runpath cannot be set
        """
        self._inner.set_runpath(set_runpath)

    def remove_runpath(self) -> None:
        """
        Remove any existing runpath from the ELF file.

        Raises:
            ElfError: If the runpath cannot be removed
        """
        self._inner.remove_runpath()

    def get_runpath(self) -> Optional[str]:
        """
        Get the current runpath from the ELF file.

        Returns:
            The current runpath, or None if no runpath is set
        """
        return self._inner.get_runpath()

    def force_rpath(self) -> None:
        """
        Force the ELF file to use DT_RPATH instead of DT_RUNPATH.

        Raises:
            ElfError: If the operation cannot be performed
        """
        self._inner.force_rpath()

    def shrink_rpath(self, rpath_prefixes: List[str]) -> None:
        """
        Shrink rpath to only include specified prefixes.

        Args:
            rpath_prefixes: List of prefixes to include in the rpath

        Raises:
            ElfError: If the rpath cannot be shrunk
        """
        self._inner.shrink_rpath(rpath_prefixes)

    def add_needed(self, needed_libs: List[str]) -> None:
        """
        Add needed libraries to the ELF file.

        Args:
            needed_libs: List of needed libraries to add

        Raises:
            ElfError: If the needed libraries cannot be added
        """
        self._inner.add_needed(needed_libs)

    def remove_needed(self, needed_libs: List[str]) -> None:
        """
        Remove needed libraries from the ELF file.

        Args:
            needed_libs: List of needed libraries to remove

        Raises:
            ElfError: If the needed libraries cannot be removed
        """
        self._inner.remove_needed(needed_libs)

    def replace_needed(self, needed_mappings: Dict[str, str]) -> None:
        """
        Replace needed libraries in the ELF file.

        Args:
            needed_mappings: Dictionary mapping old library names to new library names

        Raises:
            ElfError: If the needed libraries cannot be replaced
        """
        self._inner.replace_needed(needed_mappings)

    def get_needed(self) -> List[str]:
        """
        Get the list of needed libraries from the ELF file.

        Returns:
            List of needed libraries
        """
        return self._inner.get_needed()

    def set_interpreter(self, new_interpreter: str) -> None:
        """
        Change the interpreter path in the ELF file.

        Args:
            new_interpreter: The new interpreter path

        Raises:
            ElfError: If the interpreter cannot be changed
        """
        self._inner.set_interpreter(new_interpreter)

    def get_interpreter(self) -> Optional[str]:
        """
        Get the current interpreter from the ELF file.

        Returns:
            The current interpreter, or None if no interpreter is set
        """
        return self._inner.get_interpreter()

    def set_os_abi(self, os_abi: str) -> None:
        """
        Set the OS ABI in the ELF file.

        Args:
            os_abi: The OS ABI to set

        Raises:
            ElfError: If the OS ABI cannot be set
        """
        self._inner.set_os_abi(os_abi)

    def get_os_abi(self) -> int:
        """
        Get the OS ABI from the ELF file.

        Returns:
            The OS ABI
        """
        return self._inner.get_os_abi()

    def set_soname(self, soname: str) -> None:
        """
        Set the SONAME of the ELF file.

        Args:
            soname: The SONAME to set

        Raises:
            ElfError: If the SONAME cannot be set
        """
        self._inner.set_soname(soname)

    def get_soname(self) -> Optional[str]:
        """
        Get the SONAME of the ELF file.

        Returns:
            The SONAME, or None if no SONAME is set
        """
        return self._inner.get_soname()

    def no_default_lib(self) -> None:
        """
        Disable the default library search paths.

        Raises:
            ElfError: If the operation cannot be performed
        """
        self._inner.no_default_lib()

    def clear_version_symbol(self, symbol_name: str) -> None:
        """
        Clear a specific symbol version in the ELF file.

        Args:
            symbol_name: The name of the symbol whose version to clear

        Raises:
            ElfError: If the version cannot be cleared
        """
        self._inner.clear_version_symbol(symbol_name)

    def add_debug_tag(self) -> None:
        """
        Add a debug tag to the ELF file.

        Raises:
            ElfError: If the debug tag cannot be added
        """
        self._inner.add_debug_tag()

    def clear_exec_stack(self) -> None:
        """
        Clear the executable stack flag in the ELF file.

        Raises:
            ElfError: If the executable stack flag cannot be cleared
        """
        self._inner.clear_exec_stack()

    def set_exec_stack(self) -> None:
        """
        Set the executable stack flag in the ELF file.

        Raises:
            ElfError: If the executable stack flag cannot be set
        """
        self._inner.set_exec_stack()

    def is_exec_stack(self) -> Optional[bool]:
        """
        Get the executable stack status.

        Returns:
            True if the executable stack flag is set, False otherwise, or None if unknown
        """
        return self._inner.is_exec_stack()

    def rename_dynamic_symbols(self, symbols_map: Dict[str, str]) -> None:
        """
        Rename dynamic symbols in the ELF file.

        Args:
            symbols_map: Dictionary mapping old symbol names to new symbol names

        Raises:
            ElfError: If the symbols cannot be renamed
        """
        self._inner.rename_dynamic_symbols(symbols_map)

    def save(self, path: str) -> None:
        """
        Save the modified ELF file to a path.

        Args:
            path: Path where the modified file will be saved

        Raises:
            Exception: If the file cannot be written
        """
        self._inner.save(path)

    @classmethod
    def _from_py(cls, py_elf: PyElfContainer) -> "ElfContainer":
        elf_container = cls.__new__(cls)
        elf_container._inner = py_elf

        return elf_container
