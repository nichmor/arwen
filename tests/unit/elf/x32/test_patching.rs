use arwen::elf::ElfContainer;
// use goblin::mach::MachO;
use goblin::elf::Elf;
use rstest::rstest;
use std::path::PathBuf;

/// This test checks if the rpath of a Elf file can be set.
#[rstest]
fn test_rpath_set(#[files("tests/data/elf/x32/exec/*")] bin_path: PathBuf) {
    let data_bytes = std::fs::read(&bin_path).unwrap();

    let mut elf_container = ElfContainer::parse(&data_bytes).unwrap();

    elf_container
        .set_runpath("path_graf_path_graf_path_graf_path_graf")
        .unwrap();

    let mut changed_elf_data = Vec::new();
    elf_container.write(&mut changed_elf_data).unwrap();

    let changed_elf = Elf::parse(&changed_elf_data).unwrap();

    insta::assert_debug_snapshot!(changed_elf.runpaths);

    insta::assert_snapshot!(changed_elf_data.len());
}

/// This test checks if the rpath of a Mach-O file can be removed.
#[rstest]
fn test_rpath_remove(#[files("tests/data/elf/x32/exec/*")] bin_path: PathBuf) {
    let data_bytes = std::fs::read(&bin_path).unwrap();

    let mut elf_container = ElfContainer::parse(&data_bytes).unwrap();

    // we try to change the rpath of the file to something longer
    elf_container.remove_runpath().unwrap();

    let mut changed_elf_data = Vec::new();
    elf_container.write(&mut changed_elf_data).unwrap();

    let changed_elf = Elf::parse(&changed_elf_data).unwrap();

    insta::assert_debug_snapshot!(changed_elf.runpaths);
    insta::assert_snapshot!(changed_elf_data.len());
}

/// This test verify that a rpath can be added to a Mach-O file.
#[rstest]
fn test_rpath_add(#[files("tests/data/elf/x32/exec/*")] bin_path: PathBuf) {
    // load hello_with_rpath from data
    let data_bytes = std::fs::read(&bin_path).unwrap();

    let mut elf_container = ElfContainer::parse(&data_bytes).unwrap();

    // we try to change the rpath of the file to something longer
    elf_container
        .add_runpath("abababababababababababaabbababababababababab")
        .unwrap();

    let mut changed_elf_data = Vec::new();
    elf_container.write(&mut changed_elf_data).unwrap();

    let changed_macho = Elf::parse(&changed_elf_data).unwrap();

    insta::assert_debug_snapshot!(changed_macho.runpaths);
    insta::assert_snapshot!(changed_elf_data.len());
}

/// This test verify that a rpath can be added to a Mach-O file.
#[rstest]
fn test_force_rpath(#[files("tests/data/elf/x32/exec/*")] bin_path: PathBuf) {
    // load hello_with_rpath from data
    let data_bytes = std::fs::read(&bin_path).unwrap();

    let mut elf_container = ElfContainer::parse(&data_bytes).unwrap();

    // we try to change the rpath of the file to something longer
    elf_container.force_rpath().unwrap();

    let mut changed_elf_data = Vec::new();
    elf_container.write(&mut changed_elf_data).unwrap();

    let changed_macho = Elf::parse(&changed_elf_data).unwrap();

    insta::assert_debug_snapshot!(changed_macho.rpaths);
    insta::assert_snapshot!(changed_elf_data.len());
}

/// This test verify that a rpath can be added to a Mach-O file.
#[rstest]
fn test_set_interpreter(#[files("tests/data/elf/x32/exec/*")] bin_path: PathBuf) {
    // load hello_with_rpath from data
    let data_bytes = std::fs::read(&bin_path).unwrap();

    let mut elf_container = ElfContainer::parse(&data_bytes).unwrap();

    // we try to change the rpath of the file to something longer
    elf_container.set_interpreter("/my/ld-linux.so.2").unwrap();

    let mut changed_elf_data = Vec::new();
    elf_container.write(&mut changed_elf_data).unwrap();

    let changed_macho = Elf::parse(&changed_elf_data).unwrap();

    insta::assert_debug_snapshot!(changed_macho.interpreter);
    insta::assert_snapshot!(changed_elf_data.len());
}

/// This test verify that a rpath can be added to a Mach-O file.
#[rstest]
fn test_set_os_abi(#[files("tests/data/elf/x32/exec/*")] bin_path: PathBuf) {
    // load hello_with_rpath from data
    let data_bytes = std::fs::read(&bin_path).unwrap();

    let mut elf_container = ElfContainer::parse(&data_bytes).unwrap();

    // we try to change the rpath of the file to something longer
    elf_container.set_os_abi("freebsd").unwrap();

    let mut changed_elf_data = Vec::new();
    elf_container.write(&mut changed_elf_data).unwrap();

    let changed_macho = Elf::parse(&changed_elf_data).unwrap();

    insta::assert_debug_snapshot!(changed_macho.header);
    insta::assert_snapshot!(changed_elf_data.len());
}

/// This test verify that a rpath can be added to a Mach-O file.
#[rstest]
fn test_set_soname(#[files("tests/data/elf/x32/exec/*")] bin_path: PathBuf) {
    // load hello_with_rpath from data
    let data_bytes = std::fs::read(&bin_path).unwrap();

    let mut elf_container = ElfContainer::parse(&data_bytes).unwrap();

    // we try to change the rpath of the file to something longer
    elf_container.set_soname("my_soname").unwrap();

    let mut changed_elf_data = Vec::new();
    elf_container.write(&mut changed_elf_data).unwrap();

    let changed_macho = Elf::parse(&changed_elf_data).unwrap();

    insta::assert_debug_snapshot!(changed_macho.soname);
    insta::assert_snapshot!(changed_elf_data.len());
}

/// This test verify that a rpath can be added to a Mach-O file.
#[rstest]
fn test_add_dt_needed(#[files("tests/data/elf/x32/exec/*")] bin_path: PathBuf) {
    // load hello_with_rpath from data
    let data_bytes = std::fs::read(&bin_path).unwrap();

    let mut elf_container = ElfContainer::parse(&data_bytes).unwrap();

    // we try to change the rpath of the file to something longer
    elf_container
        .add_needed(vec!["my_dtneed".to_string()])
        .unwrap();

    let mut changed_elf_data = Vec::new();
    elf_container.write(&mut changed_elf_data).unwrap();

    let changed_macho = Elf::parse(&changed_elf_data).unwrap();

    insta::assert_debug_snapshot!(changed_macho.libraries);
    insta::assert_snapshot!(changed_elf_data.len());
}

/// This test verify that a rpath can be added to a Mach-O file.
#[rstest]
fn test_remove_dt_needed(#[files("tests/data/elf/x32/exec/*")] bin_path: PathBuf) {
    // load hello_with_rpath from data
    let data_bytes = std::fs::read(&bin_path).unwrap();

    let mut elf_container = ElfContainer::parse(&data_bytes).unwrap();

    // we try to change the rpath of the file to something longer
    elf_container
        .add_needed(vec!["my_dtneed".to_string()])
        .unwrap();

    // we try to change the rpath of the file to something longer
    elf_container
        .add_needed(vec!["my_dtneed_to_remove".to_string()])
        .unwrap();

    // we try to change the rpath of the file to something longer
    elf_container
        .remove_needed(vec!["my_dtneed_to_remove".to_string()])
        .unwrap();

    let mut changed_elf_data = Vec::new();
    elf_container.write(&mut changed_elf_data).unwrap();

    let changed_macho = Elf::parse(&changed_elf_data).unwrap();

    insta::assert_debug_snapshot!(changed_macho.libraries);
    insta::assert_snapshot!(changed_elf_data.len());
}

/// This test verify that a rpath can be added to a Mach-O file.
#[rstest]
fn test_no_default_lib(#[files("tests/data/elf/x32/exec/*")] bin_path: PathBuf) {
    // load hello_with_rpath from data
    let data_bytes = std::fs::read(&bin_path).unwrap();

    let mut elf_container = ElfContainer::parse(&data_bytes).unwrap();

    // we try to change the rpath of the file to something longer
    elf_container.no_default_lib().unwrap();

    let mut changed_elf_data = Vec::new();
    elf_container.write(&mut changed_elf_data).unwrap();

    let changed_macho = Elf::parse(&changed_elf_data).unwrap();

    insta::assert_debug_snapshot!(changed_macho.dynamic);
    insta::assert_snapshot!(changed_elf_data.len());
}

/// This test verify that a rpath can be added to a Mach-O file.
#[rstest]
fn test_add_debug_tag(#[files("tests/data/elf/x32/exec/*")] bin_path: PathBuf) {
    // load hello_with_rpath from data
    let data_bytes = std::fs::read(&bin_path).unwrap();

    let mut elf_container = ElfContainer::parse(&data_bytes).unwrap();

    // we try to change the rpath of the file to something longer
    elf_container.add_debug_tag().unwrap();

    let mut changed_elf_data = Vec::new();
    elf_container.write(&mut changed_elf_data).unwrap();

    let changed_macho = Elf::parse(&changed_elf_data).unwrap();

    insta::assert_debug_snapshot!(changed_macho.dynamic);
    insta::assert_snapshot!(changed_elf_data.len());
}

/// This test verify that a rpath can be added to a Mach-O file.
#[rstest]
fn test_set_execstack(#[files("tests/data/elf/x32/exec/*")] bin_path: PathBuf) {
    // load hello_with_rpath from data
    let data_bytes = std::fs::read(&bin_path).unwrap();

    let mut elf_container = ElfContainer::parse(&data_bytes).unwrap();

    // we try to change the rpath of the file to something longer
    elf_container.set_exec_stack().unwrap();

    let mut changed_elf_data = Vec::new();
    elf_container.write(&mut changed_elf_data).unwrap();

    let changed_macho = Elf::parse(&changed_elf_data).unwrap();

    insta::assert_debug_snapshot!(changed_macho.program_headers);
    insta::assert_snapshot!(changed_elf_data.len());
}

/// This test verify that a rpath can be added to a Mach-O file.
#[rstest]
fn test_clear_execstack(#[files("tests/data/elf/x32/exec/*")] bin_path: PathBuf) {
    // load hello_with_rpath from data
    let data_bytes = std::fs::read(&bin_path).unwrap();

    let mut elf_container = ElfContainer::parse(&data_bytes).unwrap();

    // we try to change the rpath of the file to something longer
    elf_container.set_exec_stack().unwrap();

    elf_container.clear_exec_stack().unwrap();

    let mut changed_elf_data = Vec::new();
    elf_container.write(&mut changed_elf_data).unwrap();

    let changed_macho = Elf::parse(&changed_elf_data).unwrap();

    insta::assert_debug_snapshot!(changed_macho.program_headers);
    insta::assert_snapshot!(changed_elf_data.len());
}
