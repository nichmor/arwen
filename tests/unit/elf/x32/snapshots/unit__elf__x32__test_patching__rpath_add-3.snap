---
source: tests/unit/elf/x32/test_patching.rs
expression: macho_container.data.len()
---
827540
